pkg/datapath/linux/ipsec/xfrm_collector.go

```diff
@@ -0,0 +1,277 @@
+			"Total number of transformation mode specific errors",
+			nil, map[string]string{labelErrorType: labelErrorTypeOutbound},
+		),
+		xfrmOutStateSeqErrorDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(metrics.Namespace, subsystem, "xfrm_out_state_seq_error"),
+			"Total number of sequence errors i.e. Sequence number overflow",
+			nil, map[string]string{labelErrorType: labelErrorTypeOutbound},
+		),
+		xfrmOutStateExpiredDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(metrics.Namespace, subsystem, "xfrm_out_state_expired"),
+			"Total number of expired state errors",
+			nil, map[string]string{labelErrorType: labelErrorTypeOutbound},
+		),
+		xfrmOutPolBlockDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(metrics.Namespace, subsystem, "xfrm_out_pol_block"),
+			"Total number of policy discarded errors",
+			nil, map[string]string{labelErrorType: labelErrorTypeOutbound},
+		),
+		xfrmOutPolDeadDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(metrics.Namespace, subsystem, "xfrm_out_pol_dead"),
+			"Total number of errors due to dead policy",
+			nil, map[string]string{labelErrorType: labelErrorTypeOutbound},
+		),
+		xfrmOutPolErrorDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(metrics.Namespace, subsystem, "xfrm_out_policy_error"),
+			"Total number of policy errors",
+			nil, map[string]string{labelErrorType: labelErrorTypeOutbound},
+		),
+		xfrmOutStateInvalidDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(metrics.Namespace, subsystem, "xfrm_out_state_invalid"),
+			"Total number of invalid status errors e.g. expired state",
+			nil, map[string]string{labelErrorType: labelErrorTypeOutbound},
+		),
+	}
+}
+
+func (x *xfrmCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- x.xfrmInErrorDesc
+	ch <- x.xfrmInBufferErrorDesc
+	ch <- x.xfrmInHdrErrorDesc
+	ch <- x.xfrmInNoStatesDesc
+	ch <- x.xfrmInStateProtoErrorDesc
+	ch <- x.xfrmInStateModeErrorDesc
+	ch <- x.xfrmInStateSeqErrorDesc
+	ch <- x.xfrmInStateExpiredDesc
+	ch <- x.xfrmInStateMismatchDesc
+	ch <- x.xfrmInStateInvalidDesc
+	ch <- x.xfrmInTmplMismatchDesc
+	ch <- x.xfrmInNoPolsDesc
+	ch <- x.xfrmInPolBlockDesc
+	ch <- x.xfrmInPolErrorDesc
+	ch <- x.xfrmOutErrorDesc
+	ch <- x.xfrmOutBundleGenErrorDesc
+	ch <- x.xfrmOutBundleCheckErrorDesc
+	ch <- x.xfrmOutNoStatesDesc
+	ch <- x.xfrmOutStateProtoErrorDesc
+	ch <- x.xfrmOutStateModeErrorDesc
+	ch <- x.xfrmOutStateSeqErrorDesc
+	ch <- x.xfrmOutStateExpiredDesc
+	ch <- x.xfrmOutPolBlockDesc
+	ch <- x.xfrmOutPolDeadDesc
+	ch <- x.xfrmOutPolErrorDesc
+	ch <- x.xfrmFwdHdrErrorDesc
+	ch <- x.xfrmOutStateInvalidDesc
+	ch <- x.xfrmAcquireErrorDesc
+}
+
+func (x *xfrmCollector) Collect(ch chan<- prometheus.Metric) {
+	stats, err := x.xfrmStatFunc()
+	if err != nil {
+		log.WithError(err).Error("Error while getting xfrm stats")
+		return
+	}
+
+	ch <- prometheus.MustNewConstMetric(x.xfrmInErrorDesc, prometheus.GaugeValue, float64(stats.XfrmInError))
```

Comment:
few of metrics can be combined with diff tag values, but it is not really help much in term of storage (as real number of metric is all combination of tag values), so just to simplicity, I keep it 1 to 1 mapping between metric and value in /proc/net/xfrm file.